### ISSUE TITLE
linuxdeployqt blacklists libfontconfig

### ISSUE THREAD

**Issue by boudewijnrempt** [probonopd/linuxdeployqt#261]
There's this:

tools/linuxdeployqt/shared.cpp:        excludelist << "libasound.so.2" << "libcom_err.so.2" << "libcrypt.so.1" << "libc.so.6" << "libdl.so.2" << "libdrm.so.2" << "libexpat.so.1" << "libfontconfig.so.1" << "libgcc_s.so.1" << "libgdk_pixbuf-2.0.so.0" << "libgio-2.0.so.0" << "libglib-2.0.so.0" << "libGL.so.1" << "libgobject-2.0.so.0" << "libgpg-error.so.0" << "libICE.so.6" << "libkeyutils.so.1" << "libm.so.6" << "libnsl.so.1" << "libnss3.so" << "libnssutil3.so" << "libp11-kit.so.0" << "libpangoft2-1.0.so.0" << "libpangocairo-1.0.so.0" << "libpango-1.0.so.0" << "libpthread.so.0" << "libresolv.so.2" << "librt.so.1" << "libSM.so.6" << "libstdc++.so.6" << "libusb-1.0.so.0" << "libuuid.so.1" << "libX11.so.6" << "libxcb.so.1" << "libz.so.1";

But that leads to issues like this:

https://bugs.kde.org/show_bug.cgi?id=392307

"The latest 4.0 appimage fails to start just like in bug 382944 and using QT_DEBUG_PLUGINS=1 reveals the same as in the attachment:

QLibraryPrivate::loadPlugin failed on "/tmp/.mount_appimaY4YLSz/usr/plugins/platforms/libqxcb.so" : "Cannot load library /tmp/.mount_appimaY4YLSz/usr/plugins/platforms/libqxcb.so: (/usr/lib/libfontconfig.so.1: undefined symbol: FT_Done_MM_Var)""

On ArchLinux

**Comment by probonopd** [comment]
It looks like this symbol was added only very recently, https://lists.nongnu.org/archive/html/freetype-commit/2018-01/msg00007.html

So the question is, why does anything inside your AppImage already depend on this very new symbol?

**Comment by TheAssassin** [comment]
Pretty simple: it's not a direct dependency but used by a dependency of the system's libfontconfig, hence every binary which loads libfontconfig calls that function.

**Comment by boudewijnrempt** [comment]
Okay, then it's the distribution's issue, right? this issue can be closed, and I can close my bug...

**Comment by probonopd** [comment]
Please comment on or create tickets for [libfontconfig](https://bugs.freedesktop.org/enter_bug.cgi?product=fontconfig) ([mailing list](https://lists.freedesktop.org/mailman/listinfo/fontconfig)) and [Arch](https://bugs.archlinux.org/), and describe the issue and its effects (e.g., by linking to tickets resulting from this) - before this spreads to other distributions.

* https://bugs.archlinux.org/index.php?string=FT_Done_MM_Var - no results so far
* https://bugs.freedesktop.org/buglist.cgi?quicksearch=FT_Done_MM_Var - no results so far

**Comment by TheAssassin** [comment]
Looks like the bug is coming from libfreetype2 (`FT` was already a hint in that direction). See https://bbs.archlinux.org/viewtopic.php?id=234082. libfontconfig depends on it. 

Do we bundle libfreetype2 at the moment? If yes, we might be able to solve the issue by blacklisting it.

**Comment by probonopd** [comment]
Interesting @TheAssassin, we should try that.

**Comment by TheAssassin** [comment]
Confirming that `libfreetype.so.6` is contained in AppImageUpdate's AppImage.

**Comment by TheAssassin** [comment]
See https://github.com/AppImage/AppImages/pull/323.

**Comment by TheAssassin** [comment]
Merged the PR, and restarted the linuxdeployqt build. The new exclude list should be included in the new build. Please wait a few minutes, then you can test it. If the change causes new issues that outweigh the advantages (e.g., works on Arch but not on Ubuntu), please open an issue in the AppImages repo.

**Comment by probonopd** [comment]
Now I'd be super interested in whether this fixes it for Krita....

**Comment by TheAssassin** [comment]
@probonopd simply download Krita, remove libfreetype2 and libharfbuzz, package again and upload, then someone just has to try it in Arch.

**Comment by Martchus** [comment]
Because @TheAssassin statement about Arch Linux has already gotten a :-1: I - as an Arch Linux fan - would like to comment on it, too.

> Arch is the worst "distribution" [...] **to deploy software for**.

This absolutely *correct* with emphasis on the last part. It is hard to deploy prebuilt software as a 3rd party for the distribution. Especially when this software is supposed to run under other distributions, too. However, for me is this undeniable fact is only a small flaw considering the advantages of the distribution. In fact it is so easy to create packages for it, that app images are almost useless for Arch Linux users anyways.

Additionally, I would also like to note that this is not an Arch Linux only problem. In fact, I can not run my app images currently on Tumbleweed, too. I suppose for the same reason.

To summarize:
* Users of rolling release distributions should *not* expect app images to work always on their machines. This would be expecting the impossible. In fact it is the same deal as with proprietary graphics drivers and a known issue.
* It is not an Arch Linux specific issue. Any distribution moving faster than the average is affected.

**Comment by probonopd** [comment]
Agree @Martchus, if a distribution is moving fast it's inevitable that issues are likely to pop up there first. It's also correct that the need to get applications from outside the distribution is smaller the more and the more recent applications are available inside the distribution.

If you look at an operating system as something that should provide a dependable base for third parties to develop against, then _moving targets_ are always harder than _slow-moving, long term stable_ targets. I think this is at the essence of @TheAssassin's statement. While I would never want to run a rolling release system for those reasons, I would immediately think of Arch and Tumbleweed as examples that are particular good - at what rolling release systems do.

**Comment by phw** [comment]
Just confirming that the proposed fix of excluding libharfbuzz.so.0 and libfreetype.so.6 from the AppImage works for me as well to get my own AppImage of Peek run again on Arch.

**Comment by TheAssassin** [comment]
Nice, @phw. All the feedback we receive for the exclusion of these libraries is positive. Seems like we really fixed one of the most annoying bugs we had in the last few months.
